### PR TITLE
Python Dispatcher integration with C++ dispatcher

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84966
* #84965
* __->__ #84826

Signed-off-by: Edward Z. Yang <ezyangfb.com>

From @ezyang's original PR:

There are a number of situations where we have non-backend kernels (e.g., CompositeImplicitAutograd, batching rules) which we would like to port to Python, but we have no way to integrate these ports with the overall system while using preexisting C++ registrations otherwise. This PR changes that by introducing a Python dispatcher (which can have its own kernels directly in Python), which can be interpose over ordinary C++ dispatch. The ingredients:

We introduce a new PythonDispatcher dispatch key, that has the same tenor as FuncTorchDynamicLayerFrontMode: it works by getting triggered before every other dispatch key in the dispatch key, and shunting to a Python implementation
The Python dispatcher is a per-interpreter global object that is enabled/disabled via the guard EnablePythonDispatcher/DisablePythonDispatcher. We don't make it compositional as I have no idea what a compositional version of this feature would look like. Because it is global, we don't need to memory manage it and so I use a simpler SafePyHandle (newly added) to control access to this pointer from non-Python C++. Like __torch_dispatch__, we use PyInterpreter to get to the Python interpreter to handle the dispatch.
I need to reimplement dispatch table computation logic in Python. To do this, I expose a lot more helper functions for doing computations on alias dispatch keys and similar. I also improve the pybind11 handling for DispatchKey so that you can either accept the pybind11 bound enum or a string; this simplifies our binding code. See https://github.com/pybind/pybind11/issues/483#issuecomment-1237418106 for how this works; the technique is generally useful.

I need to be able to call backend fallbacks. I do this by permitting you to call at a dispatch key which doesn't have a kernel for the operator; if the kernel doesn't exist, we check the backend fallback table instead.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by voznesenskym** [comment]
Fix after my git/ghstack mistake in #84551

**Comment by voznesenskym** [comment]
Fix after my git/ghstack mistake in #84551

File: functorch/functorch/experimental/cond.py [hunk 1]
```diff
@@ -117,21 +123,17 @@ def cond_autograd(pred, true_fn, false_fn, *operands):
...
-    def inner(pred, true_fn, false_fn, operands):
-        mode = torch._C._get_torch_dispatch_mode()
-        assert (mode is not None), "Mode should always be enabled for python fallback key"
-        with suspend_mode(mode):
-            res = trace_cond(mode, op, pred, true_fn, false_fn, operands)
-        return res
-
-    return inner
+@cond.py_impl(DispatchKey.Python)
+@cond.py_impl(ProxyTorchDispatchMode)
```
Comment (voznesenskym) [inline functorch/functorch/experimental/cond.py:127]: @ezyang thoughts on explicit mode registration? the idea is that if we see dispatch for an op, we assert that the given op is registered for the given mode. 

Should this be exclusion? Also we do not dispatch on modes yet, only keys, so we must assert that the impls match. Should we ever relax this assertion in the future?

Comment (voznesenskym) [reply functorch/functorch/experimental/cond.py:127]: Actually, I have a better idea here...

File: functorch/functorch/experimental/cond.py [hunk 2]
```diff
@@ -117,21 +123,16 @@ def cond_autograd(pred, true_fn, false_fn, *operands):
...
-cond.impl(DispatchKey.CPU, cond_dense)
-cond.impl(DispatchKey.Python, python_fallback(cond))
-cond.impl(DispatchKey.PythonTLSSnapshot, fallthrough_fn)
-cond.impl(DispatchKey.AutogradCPU, cond_autograd)
-cond.impl(DispatchKey.ADInplaceOrView, fallthrough_fn)
-cond.impl(DispatchKey.BackendSelect, fallthrough_fn)
+# TODO(voz): Make this automatic for keys, this is very ugly atm
+cond.fallthrough(DispatchKey.PythonTLSSnapshot)
+cond.fallthrough(DispatchKey.ADInplaceOrView)
+cond.fallthrough(DispatchKey.BackendSelect)
```
Comment (ezyang) [inline functorch/functorch/experimental/cond.py:138]: So, re fallthrough, I think the easiest fix will be to add C bindings to query the set of keys in c++ which are fallthrough, and then override it as necessary based on python registrations. cond.fallthrough is not symmetric with how fallthrough registrations work in C++, which is that fallthrough is a special "function" you can register.

File: torch/_dispatch/python.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+def python_dispatcher(op, ks, args, kwargs):
+    """
+    with no_python_dispatcher():
+        print(op, ks, args, kwargs)
+    """
+    k = resolve_key(op, ks.highestPriorityTypeId())
+    source = f'torch.ops.{op}.dispatch(k, *args, **kwargs)'
+    filename = f'{op}[{torch._C._dispatch_key_name(k)}]'
+    compiled = compile(source, filename, 'eval')  # TODO: maybe cache?
+    return eval(compiled, {'torch': torch, 'k': k, 'args': args, 'kwargs': kwargs})
```
Comment (ezyang) [inline torch/_dispatch/python.py:84]: This can be improved a lot, but this is an OK start

File: torch/_ops.py [hunk 4]
```diff
@@ -26,9 +32,127 @@ def dl_open_guard():
...
+
+
+class PyOperator(PyOperatorABC):
+    def __init__(self, name):
+        self._name = name
+        self.table = {}
+        self.python_key_mode_table = {}
+
+        # Make _OPNamespace not scream, this whole name based association needs a good hard look
+        self.__name__ = "pyop." + name
```
Comment (ezyang) [inline torch/_ops.py:57]: TBH, there isn't really any good reason to make pyops available at torch.ops.pyop. It is necessary for c++ bound ops as we must introduce them to Python namespace somewhere, but for Python ops you can use the normal namespacing mechanisms of Python to handle their namespacing

File: torch/_ops.py [hunk 5]
```diff
@@ -26,9 +32,127 @@ def dl_open_guard():
...
+        pyop_namespace.py_ops[name] = self
+
+    def fallthrough(self, dispatch_key):
+        self.table[dispatch_key] = self._fallthrough_fn(self, dispatch_key)
+
+    def py_impl(self, dispatch_key_or_mode):
+        def inner(fn):
+            if isinstance(dispatch_key_or_mode, type(TorchDispatchMode)):
```
Comment (ezyang) [inline torch/_ops.py:65]: don't you want issubclassof

File: torch/_ops.py [hunk 6]
```diff
@@ -26,9 +32,127 @@ def dl_open_guard():
...
+            # TODO(voz): We should walk all the nodes here / turn it into a list, topmode is ok for now.
+            curr_mode = type(torch._C._get_torch_dispatch_mode())
+            assert (
+                curr_mode is not None
+            ), "Illegal invocation of dispatch on torch._C.DispatchKey.Python without a mode."
+            assert (
+                curr_mode in self.python_key_mode_table
+            ), f"Current active mode {curr_mode} not registered"
+            # TODO(voz): The idea behind this is that we do not yet support dispatch by key + mode, only key.
+            return self.python_key_mode_table[curr_mode](*args, **kwargs)
```
Comment (ezyang) [inline torch/_ops.py:94]: Some (most?) mode implementations will need their mode object `curr_mode` as it contains essential state for doing the mode op.

File: torch/csrc/utils/python_dispatch.cpp [hunk 7]
```diff
@@ -332,46 +341,66 @@ void initDispatchBindings(PyObject* module) {
...
+#define DEF_ONE(n) .value(#n, c10::DispatchKey::n)
+
+  py::enum_<c10::DispatchKey>(m, "DispatchKey") DEF_ONE(Undefined)
+      DEF_ONE(CompositeExplicitAutogradNonFunctional)
+          DEF_ONE(CompositeExplicitAutograd)
+              DEF_ONE(CompositeImplicitAutogradNestedTensor)
+                  DEF_ONE(CompositeImplicitAutograd) DEF_ONE(AutogradOther)
+                      DEF_ONE(Autograd) DEF_ONE(BackendSelect)
+                          DEF_ONE(ADInplaceOrView) DEF_ONE(PythonTLSSnapshot)
+                              DEF_ONE(Python)
```
Comment (ezyang) [inline torch/csrc/utils/python_dispatch.cpp:374]: we should turn off clang format lol

File: torch/_ops.py (hunk 5 above)
Comment (voznesenskym) [reply torch/_ops.py:65]: Sure, that's cleaner

File: torch/_ops.py (hunk 6 above)
Comment (voznesenskym) [reply torch/_ops.py:94]: yea, but I was thinking they could query the global mode accessor - should we pass it to them?

File: torch/_ops.py (hunk 4 above)
Comment (voznesenskym) [reply torch/_ops.py:57]: This is to hack away at the first getattr. Removing this "pyop." addition gives us:

```
TypeError: '_OpNamespace' object is not callable
```
Hence the vague comment. Its a bit of a huge hack to get around this:

```
    def __getattr__(self, name):
        # Here we are creating `torch.ops.my_namespace`
        namespace = _OpNamespace(name)
        setattr(self, name, namespace)
        return namespace
```

Alternatively, we could keep a global list to check, and if the op is there, skip the namespace notion, or have a "default" namespace of sorts we consult first...

File: torch/_ops.py (hunk 5 above)
Comment (ezyang) [reply torch/_ops.py:65]: It's not just cleaner, the current code is wrong lol (in the absence of metaclass the type will always be type)

File: torch/_dispatch/python.py (hunk 3 above)
Comment (voznesenskym) [reply torch/_dispatch/python.py:84]: Let me add a comment about caching this and potentially not recreating the objects each time...

**Review by ezyang — APPROVED** [review]

File: functorch/functorch/experimental/cond.py (hunk 2 above)
Comment (voznesenskym) [reply functorch/functorch/experimental/cond.py:138]: Agreed, this one is important enough to leave an issue behind, let me do that rq

**Comment by voznesenskym** [comment]
pytorchbot  -m

File: torch/_ops.py (hunk 6 above)
Comment (ezyang) [reply torch/_ops.py:97]: when you call into the mode, that mode should already be removed from the stack. So yes we need to suspend on their behalf

File: torch/_ops.py [hunk 8]
```diff
@@ -64,7 +64,8 @@ def fallthrough(self, dispatch_key):
     def py_impl(self, dispatch_key_or_mode):
         def inner(fn):
             if inspect.isclass(dispatch_key_or_mode) and issubclass(
-                dispatch_key_or_mode, TorchDispatchMode
+<<<<<<< HEAD
```
Comment (ezyang) [inline torch/_ops.py]: yarp

**Comment by voznesenskym** [comment]
@malfet what was pushed was not what was in the PR...

**Comment by voznesenskym** [comment]
https://github.com/pytorch/pytorch/commit/35f6a69191ef762cf22b6cbfe94b8d9406e16674 differs from this state of the PR? Very strange.

**Comment by voznesenskym** [comment]
I can confirm that as it is, this branch passes dynamo tests.

**Comment by malfet** [comment]
> @malfet what was pushed was not what was in the PR...

Ok, this is 2nd change like that, so I need to write a linter into a mergebot. 
`ghstack` authored commits do not support hybrid dev mode: one can not author commit via `ghstack` and then just push changes to `head`

Will work on change ASAP that will prohibit merges to commits that do not have `ghstack-poisoned` beadcrumb in their description ( cc: @ezyang )

File: torch/csrc/utils/python_arg_parser.cpp [hunk 9]
```diff
@@ -92,6 +92,7 @@ bool should_allow_numbers_as_tensors(const std::string& name) {
       "sub",          "sub_",          "sub_out",
       "subtract",     "subtract_",     "subtract_out", // alias of sub
       "true_divide",  "true_divide_",  "true_divide_out",
+      "to",           "_to_copy",      "copy_",
```
Comment (ezyang) [inline torch/csrc/utils/python_arg_parser.cpp:95]: What's this for? This seems unrelated to this PR

**Comment by ezyang** [comment]
@malfet an alternate strategy which is more robust and IMO simpler to implement is to compare the trees of orig and head; if they are diverged, then refuse to merge.

**Comment by voznesenskym** [comment]
Closed in https://github.com/pytorch/pytorch/pull/85050

File: torch/csrc/utils/python_arg_parser.cpp (hunk 9 above)
Comment (voznesenskym) [reply torch/csrc/utils/python_arg_parser.cpp:95]: This was to get certain tests passing

File: c10/core/DispatchKeySet.h [hunk 10]
```diff
@@ -172,7 +172,9 @@ class DispatchKeySet final {
             (1ULL
              << (num_backends + static_cast<uint8_t>(toFunctionalityKey(t)) -
                  1)) -
-            1) {}
+            1) {
+    *this = add(DispatchKey::PythonDispatcher);
```
Comment (bdhirsh) [inline c10/core/DispatchKeySet.h:176]: Mostly as a call out (and reminder to myself - just spent a while trying to understand this line haha):

I think one consequence of this is that we need to be careful to use the right API's if we want to the PythonDispatcher to trigger properly during redispatches. AKA, this will properly trigger the PythonDispatcher when it's enabled:
```
// FULL_AFTER ctr will implicitly add the PythonDispatcher key to the keyset, so we can call into it
op.redispatchBoxed(ks & DispatchKeySet(DispatchKeySet::FULL_AFTER, my_dispatch_key));
```

but this redispatch will silently ignore the python dispatcher, even if it's enabled:
```
// PythonDispatcher key never gets added to the keyset, so even if we "turned it on", we'll ignore it
op.redispatchBoxed(DispatchKeySet({my_dispatch_key}));
```

I did a quick audit of `redispatch()` call-sites and I don't think there are an existing places where that happens today that are meaningful to tracing.
